src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/formula/JoinAwareWeightedUnboundVariableCount.java

```diff
@@ -0,0 +1,113 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.formula;
+
+import org.apache.jena.graph.Node;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.Join_Analyzer;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.Query_Analyzer;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Estimate selectivity of a subplan by considering
+ * - the number of new unbound variables
+ * - the position of new unbound variables (sub, pred, or obj)
+ * - number of joins contained in this subplan
+ * - types of joins contained in this subplan (star join, chain join, unusal join)
+ * To get unbound variables, the set of bindings in selectedPlans needs to be considered.
+ *
+ * This implementation is based on the formula (7) that is proposed in paper
+ * "Heuristics-based Query Reordering for Federated Queries in SPARQL 1.1 and SPARQL-LD"
+ */
+
+public class JoinAwareWeightedUnboundVariableCount {
+    protected static final double W_S = 1; // weight for subject variables
+    protected static final double W_P = 0.1; // weight for predicate variables
+    protected static final double W_O = 0.8; // weight for object variables
+
+    protected static final double J_Tc = 0.6; // weight for chain join
+    protected static final double J_Ts = 0.5; // weight for star join
+    protected static final double J_Tu = 1; // weight for unusual join
+
+    public static double estimate( final List<Query_Analyzer> selectedPlans, final Query_Analyzer subPlan ) {
+        // Add bound variables of all selected plans to a set
+        Set<Node> boundVariables = new HashSet<>();
+
+        // This part can be optimized if all formulas need to consider bound variables:
+        // adding to this set directly when adding a new subPlan to 'selectedPlans'
+        for ( Query_Analyzer plan : selectedPlans ) {
+            boundVariables = addBinds( plan.getSubs(), plan.getPreds(), plan.getObjs() );
```

Comment:
This line is overriding `boundVariables` for every `plan`. I don't think that this is what you want to do. I suggest remove the `addBinds` function and move the code from that function directly into this for loop here.